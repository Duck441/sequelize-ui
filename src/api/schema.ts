import { Schema } from '@src/core/schema'
import { versionedName } from '@src/core/utils/string'
import shortid from 'shortid'

export const SCHEMA_NOT_FOUND_ERROR = '[Schema Api Error] Schema not found'

export async function listSchemas(): Promise<Schema[]> {
  return (await get<Schema[]>(schemasKey())) || []
}

export async function getSchema(id: string): Promise<Schema> {
  const schemas = await listSchemas()
  const schema = schemas.find((s) => s.id === id)
  return schema || Promise.reject(new Error(SCHEMA_NOT_FOUND_ERROR))
}

export async function createSchema(schemaPayload: Omit<Schema, 'id'>): Promise<Schema> {
  const schemas = await listSchemas()
  const name = versionedName(
    schemaPayload.name,
    schemas.map((s) => s.name),
  )
  const id = shortid()
  const schema: Schema = { ...schemaPayload, id, name }
  await set(schemasKey(), [...schemas, schema])
  return schema
}

export async function updateSchema(schema: Schema): Promise<Schema> {
  const schemas = await listSchemas()
  const existingSchema = schemas.find((s) => s.id === schema.id)

  if (!existingSchema) {
    return Promise.reject(new Error(SCHEMA_NOT_FOUND_ERROR))
  }

  const name = versionedName(
    schema.name,
    schemas.map((s) => s.name),
  )

  const updatedSchema: Schema = removeTargetingAssociations(existingSchema, { ...schema, name })
  const updatedSchemas = schemas.map((s) => (s.id === schema.id ? updatedSchema : s))

  await set(schemasKey(), updatedSchemas)
  return updatedSchema
}

function removeTargetingAssociations(schema: Schema, updatedSchema: Schema): Schema {
  const removedModelIds = schema.models
    .filter((m) => !updatedSchema.models.some((um) => um.id === m.id))
    .map((m) => m.id)

  const models = updatedSchema.models.map((m) => ({
    ...m,
    associations: m.associations.filter((a) => !removedModelIds.includes(a.targetModelId)),
  }))

  return { ...updatedSchema, models }
}

export async function deleteSchema(id: string): Promise<void> {
  const schemas = await listSchemas()
  if (schemas.length) {
    const updatedSchemas = schemas.filter((s) => s.id !== id)
    await set(schemasKey(), updatedSchemas)
  }
}

export async function clearSchemas(): Promise<void> {
  return await remove(schemasKey())
}

function get<T>(key: string): Promise<T | null> {
  return new Promise<T | null>((resolve, reject) => {
    try {
      const item: string | null = localStorage.getItem(lsKey(key))
      if (item === null) return resolve(item)
      const result: T = JSON.parse(item)
      resolve(result)
    } catch (e) {
      reject(e)
    }
  })
}

function set<T>(key: string, value: T): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      const payload = JSON.stringify(value)
      localStorage.setItem(lsKey(key), payload)
      resolve()
    } catch (e) {
      reject(e)
    }
  })
}

function remove(key: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      localStorage.removeItem(lsKey(key))
      resolve()
    } catch (e) {
      reject(e)
    }
  })
}

function schemasKey(): string {
  return '/schemas'
}

const NAMESPACE = `__SEQUELIZEUI__`
function lsKey(key: string): string {
  return NAMESPACE + key
}
