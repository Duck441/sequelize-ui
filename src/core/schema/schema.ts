import { now } from '@src/utils/dateTime'
import shortid from 'shortid'
import { Association, AssociationTypeType } from './association'
import { DataType, DataTypeType } from './dataType'

export type Schema = {
  id: string
  name: string
  models: Model[]
  createdAt: string
  updatedAt: string
}

export type Model = {
  id: string
  name: string
  fields: Field[]
  associations: Association[]
  createdAt: string
  updatedAt: string
}

export type Field = {
  id: string
  name: string
  type: DataType
  primaryKey?: boolean
  required?: boolean
  unique?: boolean
  generated?: boolean
}

export function emptySchema(): Schema {
  const time = now()
  return {
    id: '',
    name: '',
    models: [],
    createdAt: time,
    updatedAt: time,
  }
}

export function emptyModel(): Model {
  const time = now()
  return {
    id: shortid(),
    name: '',
    fields: [],
    associations: [],
    createdAt: time,
    updatedAt: time,
  }
}

export function emptyField(): Field {
  return { id: shortid(), name: '', type: { type: DataTypeType.String } }
}

export function emptyAssociation(
  sourceModelId: Model['id'],
  targetModelId: Model['id'],
): Association {
  return {
    id: shortid(),
    sourceModelId,
    type: { type: AssociationTypeType.BelongsTo },
    targetModelId,
  }
}
