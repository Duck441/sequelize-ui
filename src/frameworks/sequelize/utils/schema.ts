import { DbOptions } from '@src/core/database'
import { Field, Model, Schema } from '@src/core/schema'
import { idField, prefixPk } from './field'

type NormalizeSchemaArgs = {
  schema: Schema
  dbOptions: DbOptions
}
export const normalizeSchema = ({ schema, dbOptions }: NormalizeSchemaArgs): Schema => {
  return { ...schema, models: schema.models.map((model) => normalizeModel({ model, dbOptions })) }
}

type NormalizeModelArgs = {
  model: Model
  dbOptions: DbOptions
}
const normalizeModel = ({ model, dbOptions }: NormalizeModelArgs): Model => {
  return { ...model, fields: getFieldsWithPk({ model, dbOptions }) }
}

type GetFieldsWithIdArgs = {
  model: Model
  dbOptions: DbOptions
}
function getFieldsWithPk({ model, dbOptions }: GetFieldsWithIdArgs): Field[] {
  const pks = model.fields.filter((f) => f.primaryKey)

  // Don't apply any prefixes if pk is composite
  if (pks.length > 1) return model.fields
  // Add explicit pk field for TypeScript classes
  if (pks.length === 0) return [idField({ model, dbOptions }), ...model.fields]
  // Prefix pk with model name if pk is standard  format and prefixPk option is true
  return model.fields.map((field) => prefixPk({ field, model, dbOptions }))
}
