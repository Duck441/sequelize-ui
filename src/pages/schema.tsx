import {
  DatabaseCaseStyle,
  DatabaseNounForm,
  DatabaseOptions,
  SqlDialect,
} from '@src/core/database'
import DbOptionsForm from '@src/ui/components/DbOptionsForm'
import Layout from '@src/ui/components/Layout'
import ModelModule from '@src/ui/components/ModelModule'
import SchemaModule from '@src/ui/components/SchemaModule'
import useSchemaState, { SchemaEditStateType } from '@src/ui/hooks/useEditSchema'
import useGeneratedCode from '@src/ui/hooks/useGeneratedCode'
import dynamic from 'next/dynamic'
import React, { useCallback, useState } from 'react'

const CodeViewer = dynamic(() => import('@src/ui/components/CodeViewer'))

function SchemaPage(): React.ReactElement {
  return (
    <Layout title="Schema | Sequelize UI">
      <SchemaPageContent />
    </Layout>
  )
}

const defaultDbOptions: DatabaseOptions = {
  sqlDialect: SqlDialect.Postgres,
  caseStyle: DatabaseCaseStyle.Snake,
  nounForm: DatabaseNounForm.Singular,
  timestamps: true,
}

function SchemaPageContent(): React.ReactElement {
  const {
    schema,
    schemas,
    editState,
    error,
    edit,
    update,
    destroy,
    addModel,
    editModel,
    updateModel,
    deleteModel,
    cancel,
  } = useSchemaState()

  const [dbOptions, setDbOptions] = useState<DatabaseOptions>(defaultDbOptions)
  const { root } = useGeneratedCode({ schema, dbOptions })
  const [viewCode, setViewCode] = useState<boolean>(false)
  const handleClickViewCode = useCallback(() => setViewCode((x) => !x), [])

  if (error) return <p>{error}</p>
  if (schema === undefined) return <p>Loading Schemas</p>

  return (
    <>
      <button onClick={handleClickViewCode}>{viewCode ? 'Hide Code' : 'View Code'}</button>
      <SchemaModule
        schema={schema}
        schemas={schemas || []}
        editing={editState.type === SchemaEditStateType.EditingSchema}
        onEdit={edit}
        onUpdate={update}
        onCancel={cancel}
      />

      <button onClick={destroy}>Delete</button>
      {viewCode && root && (
        <>
          <DbOptionsForm dbOptions={dbOptions} onChange={setDbOptions} />
          <CodeViewer cacheKey={schema.id} root={root} />
        </>
      )}
      <h3>Models</h3>
      <button type="button" onClick={addModel}>
        Add model
      </button>
      {schema.models.length ? (
        <ul>
          {schema.models.map((model) => (
            <ModelModule
              key={`model-module-${model.id}`}
              model={model}
              schema={schema}
              editing={
                editState.type === SchemaEditStateType.EditingModel && editState.id === model.id
              }
              disabled={
                editState.type === SchemaEditStateType.EditingSchema ||
                (editState.type === SchemaEditStateType.EditingModel && editState.id !== model.id)
              }
              onRequestEdit={editModel}
              onRequestDelete={deleteModel}
              onRequestCancel={cancel}
              onChange={updateModel}
            />
          ))}
        </ul>
      ) : null}
    </>
  )
}

export default SchemaPage
