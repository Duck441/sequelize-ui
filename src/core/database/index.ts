import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: DatabaseCaseStyle
  nounForm: DatabaseNounForm
}

export enum DatabaseCaseStyle {
  Snake = 'snake',
  Camel = 'camel',
}

export enum DatabaseNounForm {
  Singular = 'singular',
  Plural = 'plural',
}

export enum SqlDialect {
  MariaDb = 'mariadb',
  MsSql = 'mssql',
  MySql = 'mysql',
  Postgres = 'postgres',
  Sqlite = 'sqlite',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MariaDB'
    case SqlDialect.MsSql:
      return 'SQL Server'
    case SqlDialect.MySql:
      return 'MySQL'
    case SqlDialect.Postgres:
      return 'PostgreSQL'
    case SqlDialect.Sqlite:
      return 'SQLite'
  }
}

export const sqlDialectEnvVar = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MARIADB'
    case SqlDialect.MsSql:
      return 'MSSQL'
    case SqlDialect.MySql:
      return 'MYSQL'
    case SqlDialect.Postgres:
      return 'POSTGRES'
    case SqlDialect.Sqlite:
      return 'SQLITE'
  }
}

export const parseSqlDialect = (dialect: string): SqlDialect | null => {
  switch (dialect) {
    case 'mariadb':
      return SqlDialect.MariaDb
    case 'mssql':
      return SqlDialect.MsSql
    case 'mysql':
      return SqlDialect.MySql
    case 'postgres':
      return SqlDialect.Postgres
    case 'sqlite':
      return SqlDialect.Sqlite
    default:
      return null
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectPassword = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'Password1'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectHost = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'localhost'
    case SqlDialect.MsSql:
      return 'localhost'
    case SqlDialect.MySql:
      return 'localhost'
    case SqlDialect.Postgres:
      return 'localhost'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectStorage = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return null
    case SqlDialect.MsSql:
      return null
    case SqlDialect.MySql:
      return null
    case SqlDialect.Postgres:
      return null
    case SqlDialect.Sqlite:
      return '.tmp/data.db'
  }
}

export const displayDatabaseNounForm = (nounForm: DatabaseNounForm): string => {
  switch (nounForm) {
    case DatabaseNounForm.Singular:
      return 'singular'
    case DatabaseNounForm.Plural:
      return 'plural'
  }
}

export const displayDatabaseCaseStyle = (caseStyle: DatabaseCaseStyle): string => {
  switch (caseStyle) {
    case DatabaseCaseStyle.Snake:
      return 'snake'
    case DatabaseCaseStyle.Camel:
      return 'camel'
  }
}

export const defaultDbOptions: DatabaseOptions = {
  sqlDialect: SqlDialect.Postgres,
  timestamps: true,
  caseStyle: DatabaseCaseStyle.Snake,
  nounForm: DatabaseNounForm.Plural,
}

export const MAX_IDENTIFIER_LENGTH = 63
