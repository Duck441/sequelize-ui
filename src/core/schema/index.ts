export * from './association'
export * from './dataType'
export * from './schema'
export * from './validation'
