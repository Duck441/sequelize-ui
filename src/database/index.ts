import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: 'snake' | 'camel'
  nounForm: 'singular' | 'plural'
}

export enum SqlDialect {
  MariaDb = 'MARIADB',
  MsSql = 'MSSQL',
  MySql = 'MYSQL',
  Postgres = 'POSTGRES',
  Sqlite = 'SQLITE',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'mariadb'
    case SqlDialect.MsSql:
      return 'mssql'
    case SqlDialect.MySql:
      return 'mysql'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return 'sqlite'
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}
