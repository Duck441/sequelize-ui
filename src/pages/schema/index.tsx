import { deleteSchema, getSchema, updateSchema } from '@src/api/schema'
import { Schema } from '@src/core/schema'
import { goTo } from '@src/routing/navigation'
import { indexRoute, parseSchemaId } from '@src/routing/routes'
import Flyout from '@src/ui/components/Flyout'
import SchemaFlyout from '@src/ui/components/SchemaFlyout'
import useAsync from '@src/ui/hooks/useAsync'
import { useRouter } from 'next/router'
import React from 'react'

export default function SchemaPage(): React.ReactElement {
  const [schema, setSchema] = React.useState<Schema>()
  const router = useRouter()
  // TODO: Navigate home with error message when no id
  const schemaId = parseSchemaId(router.query)

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const getData = React.useCallback(() => getSchema(schemaId!), [schemaId])

  const { refetch } = useAsync({ getData, skip: !schemaId, onLoad: setSchema })

  const handleChange = async (schema: Schema) => {
    const updated = await updateSchema(schema)
    setSchema(updated)
    refetch()
    return updated
  }

  const handleDelete = async () => {
    schema && (await deleteSchema(schema.id))
    goTo(indexRoute())
  }

  const handleCancel = () => goTo(indexRoute())

  return schema ? (
    <SchemaFlyout
      schema={schema}
      schemas={[]}
      onChange={handleChange}
      onDelete={handleDelete}
      onClickClose={handleCancel}
    />
  ) : (
    <Flyout title="" onClickClose={handleCancel} controls={<></>} />
  )
}
