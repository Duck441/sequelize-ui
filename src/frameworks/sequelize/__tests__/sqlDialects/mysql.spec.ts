import { defaultDbOptions, SqlDialect } from '@src/core/database'
import schema from '@src/frameworks/__fixtures__/schemas/dataTypes'
import { SequelizeFramework } from '../..'

describe('Sequelize Framework ', () => {
  describe('MySQL', () => {
    it('generates correct code with supported types and the proper config', () => {
      const code = SequelizeFramework.generate({
        schema,
        dbOptions: { ...defaultDbOptions, sqlDialect: SqlDialect.MySql },
      })
      expect(code).toMatchSnapshot()
    })
  })
})
