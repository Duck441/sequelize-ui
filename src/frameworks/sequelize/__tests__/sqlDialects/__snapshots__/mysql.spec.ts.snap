// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequelize Framework  MySQL generates correct code with supported types and the proper config 1`] = `
Object {
  "files": Array [
    Object {
      "files": Array [
        Object {
          "content": "import type { Sequelize, Model } from 'sequelize'
import { Category } from './Category'
import type { CategoryAttributes, CategoryCreationAttributes } from './Category'

export {
  Category
}

export type {
  CategoryAttributes,
  CategoryCreationAttributes
}

export function initModels(sequelize: Sequelize) {
  Category.initModel(sequelize)

  

  return {
    Category
  }
}
",
          "name": "index.ts",
          "type": "FILE",
        },
        Object {
          "content": "import Sequelize, { DataTypes, Model, Optional } from 'sequelize'
import { Json } from '../types'

export interface CategoryAttributes {
  string?: string
  stringWithLength?: string
  text?: string
  //// DataTypes.CITEXT not supported for MySQL
  // ciText?: string
  integer?: number
  integerUnsigned?: number
  integerAutoincrement?: number
  bigInt?: number
  smallInt?: number
  float?: number
  real?: number
  double?: number
  decimal?: number
  decimalPrecision?: number
  decimalPrecisionAndScale?: number
  dateTime?: string
  dateTimeDefaultNow?: string
  date?: string
  dateDefaultNow?: string
  time?: string
  timeDefaultNow?: string
  boolean?: boolean
  enum?: 
  //// DataTypes.ARRAY(DataTypes.STRING) not supported for MySQL
  // array?: string[]
  json?: Json
  //// DataTypes.JSONB not supported for MySQL
  // jsonB?: Json
  blob?: Buffer
  uuid?: string
  uuidDefaultV4?: string
  uuidDefaultV1?: string
}

export type CategoryId = number
export type CategoryCreationAttributes = CategoryAttributes & { id?: number }

export class Category extends Model<CategoryAttributes, CategoryCreationAttributes> implements CategoryAttributes {
  public readonly id?: number
  public string?: string
  public stringWithLength?: string
  public text?: string
  //// DataTypes.CITEXT not supported for MySQL
  // public ciText?: string
  public integer?: number
  public integerUnsigned?: number
  public integerAutoincrement?: number
  public bigInt?: number
  public smallInt?: number
  public float?: number
  public real?: number
  public double?: number
  public decimal?: number
  public decimalPrecision?: number
  public decimalPrecisionAndScale?: number
  public dateTime?: string
  public dateTimeDefaultNow?: string
  public date?: string
  public dateDefaultNow?: string
  public time?: string
  public timeDefaultNow?: string
  public boolean?: boolean
  public enum?: 
  //// DataTypes.ARRAY(DataTypes.STRING) not supported for MySQL
  // public array?: string[]
  public json?: Json
  //// DataTypes.JSONB not supported for MySQL
  // public jsonB?: Json
  public blob?: Buffer
  public uuid?: string
  public uuidDefaultV4?: string
  public uuidDefaultV1?: string
  
  static initModel(sequelize: Sequelize.Sequelize): typeof Category {
    Category.init({
      string: {
        type: DataTypes.STRING
      },
      stringWithLength: {
        type: DataTypes.STRING(100)
      },
      text: {
        type: DataTypes.TEXT
      },
      //// DataTypes.CITEXT not supported for MySQL
      // ciText: {
      //   type: DataTypes.CITEXT
      // },
      integer: {
        type: DataTypes.INTEGER
      },
      integerUnsigned: {
        type: DataTypes.INTEGER.UNSIGNED
      },
      integerAutoincrement: {
        type: DataTypes.INTEGER,
        autoIncrement: true
      },
      bigInt: {
        type: DataTypes.BIGINT
      },
      smallInt: {
        type: DataTypes.SMALLINT
      },
      float: {
        type: DataTypes.FLOAT
      },
      real: {
        type: DataTypes.REAL
      },
      double: {
        type: DataTypes.DOUBLE
      },
      decimal: {
        type: DataTypes.DECIMAL
      },
      decimalPrecision: {
        type: DataTypes.DECIMAL.PRECISION(14)
      },
      decimalPrecisionAndScale: {
        type: DataTypes.DECIMAL.PRECISION(14, 2)
      },
      dateTime: {
        type: DataTypes.DATE
      },
      dateTimeDefaultNow: {
        type: DataTypes.DATE,
        defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
      },
      date: {
        type: DataTypes.DATEONLY
      },
      dateDefaultNow: {
        type: DataTypes.DATEONLY,
        defaultValue: Sequelize.literal('CURRENT_DATE')
      },
      time: {
        type: DataTypes.TIME
      },
      timeDefaultNow: {
        type: DataTypes.TIME,
        defaultValue: Sequelize.literal('CURRENT_TIME')
      },
      boolean: {
        type: DataTypes.BOOLEAN
      },
      enum: {
        type: DataTypes.ENUM()
      },
      //// DataTypes.ARRAY(DataTypes.STRING) not supported for MySQL
      // array: {
      //   type: DataTypes.ARRAY(DataTypes.STRING)
      // },
      json: {
        type: DataTypes.JSON
      },
      //// DataTypes.JSONB not supported for MySQL
      // jsonB: {
      //   type: DataTypes.JSONB
      // },
      blob: {
        type: DataTypes.BLOB
      },
      uuid: {
        type: DataTypes.UUID
      },
      uuidDefaultV4: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4
      },
      uuidDefaultV1: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV1
      }
    }, {
      sequelize
    })
    
    return Category
  }
}
",
          "name": "Category.ts",
          "type": "FILE",
        },
      ],
      "name": "models",
      "type": "DIRECTORY",
    },
    Object {
      "content": "node_modules/
dist
npm-debug.log*
yarn-debug.log*
yarn-error.log*
.npm
.tmp
",
      "name": ".gitignore",
      "type": "FILE",
    },
    Object {
      "content": "import { Sequelize } from 'sequelize'

const db: Sequelize = new Sequelize({
  dialect: 'mysql',
  database: process.env.MYSQL_DB_NAME || 'data_types',
  username: process.env.MYSQL_DB_USERNAME || 'root',
  password: process.env.MYSQL_DB_PASSWORD || 'root',
  host: process.env.MYSQL_DB_HOST || 'localhost',
  port: parseInt(process.env.MYSQL_DB_PORT || '3306'),
  define: {
    underscored: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  }
})

export default db
",
      "name": "db.ts",
      "type": "FILE",
    },
    Object {
      "content": "{
  \\"name\\": \\"data-types\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"\\",
  \\"main\\": \\"server.ts\\",
  \\"scripts\\": {
    \\"test\\": \\"echo \\\\\\"Error: no test specified\\\\\\" && exit 1\\",
    \\"build\\": \\"tsc\\",
    \\"start\\": \\"node ./dist/server.js\\"
  },
  \\"author\\": \\"\\",
  \\"license\\": \\"ISC\\",
  \\"dependencies\\": {
    \\"mysql2\\": \\"^2.2.5\\",
    \\"sequelize\\": \\"^6.3.5\\"
  },
  \\"devDependencies\\": {
    \\"@types/node\\": \\"^14.14.20\\",
    \\"@types/validator\\": \\"^13.1.3\\",
    \\"typescript\\": \\"^4.1.3\\"
  }
}

",
      "name": "package.json",
      "type": "FILE",
    },
    Object {
      "content": "import db from './db'
import { initModels } from './models'

initModels(db)

db.sync({ force: true })
",
      "name": "server.ts",
      "type": "FILE",
    },
    Object {
      "content": "export interface JsonMap {[member: string]: string | number | boolean | null | JsonArray | JsonMap }

export interface JsonArray extends Array<string | number | boolean | null | JsonArray | JsonMap> {}

export type Json = JsonMap | JsonArray | string | number | boolean | null
",
      "name": "types.ts",
      "type": "FILE",
    },
    Object {
      "content": "{
    \\"compilerOptions\\": {
      \\"target\\": \\"ES2020\\",
      \\"module\\": \\"commonjs\\",
      \\"lib\\": [\\"ES2020\\"],
      \\"outDir\\": \\"dist\\",
      \\"strict\\": true,
      \\"moduleResolution\\": \\"node\\",
      \\"esModuleInterop\\": true,
      \\"forceConsistentCasingInFileNames\\": true
    }
  }
",
      "name": "tsconfig.json",
      "type": "FILE",
    },
  ],
  "name": "data-types",
  "type": "DIRECTORY",
}
`;
